[concept entry] Dart map putIfAbsent function (#4405)
Codecademy/docs: content/dart/concepts/map/terms/putifabsent/putifabsent.md

---
Title: '.putIfAbsent()'
Description: 'Inserts a key-value pair into a map if the key is not already present.'
Subjects:
  - 'Code Foundations'
  - 'Computer Science'
Tags:
  - 'Dart'
  - 'map'
CatalogContent:
  - 'learn-dart'
  - 'paths/computer-science'
---

The **`.putifabsent()`** method in Dart is used to insert a key-value pair into a map if the key is not already present. If the key is already present, it does nothing and returns the existing value associated with the key.

## Syntax

```pseudo
V putIfAbsent(K key, V ifAbsent())
```

- `key`: The key to check for or add to the map.
- `ifAbsent`: A function that returns the value associated with the key if it is not already in the map.

## Example

> In this example, the key 'key3' is not present in the map, so the ifAbsent function () => 3 is invoked to provide the value 3. The key 'key1' and 'key2' are present so the ifAbsent function is not invoked and the existing values are returned.

Reviewer: ```suggestion
In this example, the key `key3` is not present in the map, so the `ifAbsent()` function `() => 3` is invoked to provide the value 3. The keys `key1` and `key2` are present so the `ifAbsent()` function is not invoked and the existing values are returned.
```

```dart

> void main() {
>   var map = {'key1': 1, 'key2': 2};
>   // 'key3' is not present, so the value 3 will be provided by the ifAbsent function
>   map.putIfAbsent('key3', () => 3); // {'key1': 1, 'key2': 2, 'key3': 3}
>   // 'key1' is present, so the ifAbsent function is not invoked
>   map.putIfAbsent('key1', () => 5); // Notice the output stays the same {'key1': 1, 'key2': 2, 'key3': 3}
>   print(map); // output {'key1': 1, 'key2': 2, 'key3': 3}
> }

Reviewer: ```suggestion
void main() {
  var map = {'key1': 1, 'key2': 2};
  // 'key3' is not present, so the value 3 will be provided by the ifAbsent function
  map.putIfAbsent('key3', () => 3); // {'key1': 1, 'key2': 2, 'key3': 3}
  // 'key1' is present, so the ifAbsent function is not invoked
  print(map.putIfAbsent('key1', () => 5)); // Notice the output is the existing value associated with key1
  print(map); // output {'key1': 1, 'key2': 2, 'key3': 3}
}
```

It makes sense to show how the method returns the existing value of the key.

```

Here is the output for the above code:

```shell

> {key1: 1, key2: 2, key3: 3}

Reviewer: ```suggestion
1
{key1: 1, key2: 2, key3: 3}
```

---

> Title: '.putifabsent()'

Reviewer 2: ```suggestion
Title: '.putIfAbsent()'
```

> Description: '.putifabsent() is used to insert a key-value pair into a map if the key is not already present. '

Reviewer 2: ```suggestion
Description: 'Inserts a key-value pair into a map if the key is not already present.'
```

Subjects:
  - 'Code Foundations'
  - 'Computer Science'
Tags:
  - 'Dart'
  - 'map'
CatalogContent:
  - 'learn-dart'
  - 'paths/computer-science'
---

> The .putifabsent() method in Dart is used to insert a key-value pair into a map if the key is not already present. If the key is already present, it does nothing and returns the existing value associated with the key.

Reviewer 2: ```suggestion
The **`.putifabsent()`** method in Dart is used to insert a key-value pair into a map if the key is not already present. If the key is already present, it does nothing and returns the existing value associated with the key.
```

## Syntax

```pseudo
V putIfAbsent(K key, V ifAbsent())
```

> - `Key`: The key to check for or add in the map.
> - `ifabsent`: A function that returns the value to be associated with the key if it is not already present in the map.

Reviewer 2: ```suggestion
- `key`: The key to check for or add to the map.
- `ifAbsent`: A function that returns the value associated with the key if it is not already in the map.
```

## Example

In this example, the key 'key3' is not present in the map, so the ifAbsent function () => 3 is invoked to provide the value 3. The key 'key1' and 'key2' are present so the ifAbsent function is not invoked and the existing values are returned.

```dart

> void main() {
>   var map = {'key1': 1, 'key2': 2};
> // 'key3' is not present, so the value 3 will be provided by the ifabsent function
> map.putifabsent('key3', () => 3); // {'key1': 1, 'key2': 2, 'key3': 3};
> 
>   // 'key1' is present, so the ifabsent function is not invoked
> map.putifabsent('key1', () => 5); // Notice the output stays the same {'key1': 1, 'key2': 2, 'key3': 3};
> 
> print(map); // output {'key1': 1, 'key2': 2, 'key3': 3}
>  

Reviewer 2: ```suggestion
void main() {
  var map = {'key1': 1, 'key2': 2};
  // 'key3' is not present, so the value 3 will be provided by the ifAbsent function
  map.putIfAbsent('key3', () => 3); // {'key1': 1, 'key2': 2, 'key3': 3}
  // 'key1' is present, so the ifAbsent function is not invoked
  map.putIfAbsent('key1', () => 5); // Notice the output stays the same {'key1': 1, 'key2': 2, 'key3': 3}
  print(map); // output {'key1': 1, 'key2': 2, 'key3': 3}
}
```
The code had a missing bracket and the method must be called by its exact name as it is case-sensitive.

```

Here is the output for the above code:

> ```

Reviewer 2: ```suggestion
```shell
```
The output block should be inside ` ```shell`.

> {'key1': 1, 'key2': 2, 'key3': 3}

Reviewer 2: ```suggestion
{key1: 1, key2: 2, key3: 3}
```
